core/drivers/sunxi_uart.c

```diff
@@ -55,28 +54,51 @@
 #define UART_REG_USR_RFNE (0x1 << 0x3)
 #define UART_REG_USR_RFF  (0x1 << 0x4)
 
-void sunxi_uart_init(vaddr_t __unused base)
+static vaddr_t base_addr(struct sunxi_uart_data *pd)
 {
-	/* do nothing, debug uart(uart0) share with normal world,
-	 * everything for uart0 is ready now.
-	 */
+	return cpu_mmu_enabled() ? pd->vbase : pd->pbase;
```

Comment:
We could also check `pbase`, because we don't expect physical address zero. What do you think of:
```C
static vaddr_t base_addr(struct pl011_data *pd)
{
	vaddr_t base = cpu_mmu_enabled() ? pd->vbase : pd->pbase;

	assert(base);
	return base;
}
```